train.py

```diff
@@ -0,0 +1,200 @@
+
+    device = torch.device(cfg.train.device)
+
+    # model
+    unet = Unet(
+        in_channels=cfg.models.unet.in_out_channels,
+        out_channels=cfg.models.unet.in_out_channels,
+        dim=cfg.models.unet.dim,
+        dim_mults=cfg.models.unet.dim_mults,
+        kernel_size=cfg.models.unet.kernel_size,
+        resnet_block_groups=cfg.models.unet.num_resnet_groups,
+    ).to(device)
+
+    # forward diffusion
+    forward_diffusion = ForwardDiffusion(
+        beta_start=cfg.models.forward_diffusion.beta_start,
+        beta_end=cfg.models.forward_diffusion.beta_end,
+        timesteps=cfg.models.forward_diffusion.timesteps,
+        schedule_type=cfg.models.forward_diffusion.schedule_type,
+    ).to(device)
+
+    # setting up optimizer
+    optimizer = optim.AdamW(unet.parameters(), lr=cfg.train.lr, weight_decay=cfg.train.weight_decay)
+
+    # load checkpoint if specified in cfg
+    if cfg.paths.load_ckpt_path is not None:
+        checkpoint = torch.load(cfg.paths.load_ckpt_path)
+
+        unet.load_state_dict(checkpoint["model"])
+        forward_diffusion.load_state_dict(checkpoint["forward_diffusion"])
+        optimizer.load_state_dict(checkpoint["optimizer"])
+
+    # ckpt specifies directory and name of the file is name of the experiment in wandb
+    save_path = f"{cfg.paths.save_ckpt_dir}/{cfg.logger.run_name}.ckpt"
+
+    # step counts for logging to wandb
+    train_step_count = 0
+    val_step_count = 0
+
+    for epoch in range(cfg.train.num_epochs):
+        # train epoch
+        train_loop = tqdm(enumerate(train_dataloader), total=len(train_dataloader))
+
+        for batch_idx, batch in train_loop:
+            # metrics returns loss and additional metrics if specified in step function
+            metrics = step(unet, forward_diffusion, batch, device, split="train")
+
+            loss = metrics["train/loss"]
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            train_loop.set_postfix(loss=loss.item())
+
+            train_step_count += 1
+
+            if (batch_idx + 1) % cfg.logger.log_every_n_steps == 0:
+                # log metrics
+                wandb.log(metrics, step=train_step_count)
+
+                # save model and optimizer states
+                save_checkpoint(unet, forward_diffusion, optimizer, cfg, save_path=save_path)
+
+        # val epoch
+        val_loop = tqdm(enumerate(val_dataloader), total=len(val_dataloader))
+
+        for batch_idx, batch in val_loop:
+            # metrics returns loss and additional metrics if specified in step function
+            metrics = step(unet, forward_diffusion, batch, device, split="val")
+
+            loss = metrics["val/loss"]
+
+            val_loop.set_postfix(loss=loss.item())
+
+            val_step_count += 1
+
+            if (batch_idx + 1) % cfg.logger.log_every_n_steps == 0:
+                # log metrics
+                wandb.log(metrics, step=val_step_count)
```

Comment:
For validation it would be better to get the metrics averaged over the whole dataset, rather than for just some batches.

I think you can also use the `train_step_count` here, so the values will be aligned with the train stats. Alternatively you can also do `step=epoch`.